pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,500 @@
+			}
+		}
+	}
+
+	if !deleted {
+		c.statusCollectorNameToSpec[statusCollector.Name] = &statusCollector.Spec // readonly
+	} else {
+		delete(c.statusCollectorNameToSpec, statusCollector.Name)
+	}
+
+	return combinedStatusIdentifiersToQueue
+}
+
+// ResolutionExists returns true if a combinedstatus resolution exists for
+// the given binding name and workload object identifier.
+func (c *combinedStatusResolver) ResolutionExists(bindingName string, objectIdentifier util.ObjectIdentifier) bool {
+	c.RLock()
+	defer c.RUnlock()
+
+	if resolutions, exists := c.bindingNameToResolutions[bindingName]; exists {
+		_, exists = resolutions[objectIdentifier]
+		return exists
+	}
+
+	return false
+}
+
+// fetchMissingStatusCollectorSpecs fetches the missing statuscollector specs
+// from the given lister and updates the cache.
+// The method is expected to be called with the write lock held.
+func (c *combinedStatusResolver) fetchMissingStatusCollectorSpecsLocked(statusCollectorLister controllisters.StatusCollectorLister,
+	statusCollectorNames []string) {
+	for _, statusCollectorName := range statusCollectorNames {
+		if _, exists := c.statusCollectorNameToSpec[statusCollectorName]; exists {
+			continue // this method is not responsible for keeping the cache up-to-date
+		}
+
+		statusCollector, err := statusCollectorLister.Get(statusCollectorName)
+		if err != nil {
+			// fetch error should not disturb the flow.
+			// a missing spec will be reconciled when the status collector is created/updated.
+			runtime2.HandleError(fmt.Errorf("failed to get statuscollector %s: %w", statusCollectorName, err))
+			return
+		}
+
+		c.statusCollectorNameToSpec[statusCollectorName] = &statusCollector.Spec // readonly
+	}
+}
+
+// evaluateWorkStatusesPerBindingReadLocked evaluates workstatuses associated
+// with the given workload identifiers and destinations.
+// The returned set contains the identifiers of combinedstatus objects that
+// should be queued for syncing.
+// The method is expected to be called with the read lock held.
+func (c *combinedStatusResolver) evaluateWorkStatusesPerBindingReadLocked(bindingName string,
+	workloadObjIdentifiersToEvaluate sets.Set[util.ObjectIdentifier], destinations sets.Set[string],
+	workStatusIndexer cache.Indexer) sets.Set[util.ObjectIdentifier] {
+	workloadObjectToError := make(map[util.ObjectIdentifier]error) // TODO: use
+	combinedStatusesToQueue := sets.Set[util.ObjectIdentifier]{}
+
+	for workloadObjIdentifier := range workloadObjIdentifiersToEvaluate {
+		for destination := range destinations {
+			// fetch workstatus
+			indexKey := util.KeyFromSourceRefAndWecName(util.SourceRefFromObjectIdentifier(workloadObjIdentifier),
+				destination)
+
+			objs, err := workStatusIndexer.ByIndex(workStatusIdentificationIndexKey, indexKey) // one obj expected
+			if err != nil {
+				workloadObjectToError[workloadObjIdentifier] = err
+				continue
+			}
+
+			if len(objs) == 0 {
+				continue
```

Comment:
Another "no change". Shouldn't csResolution be deleted here?

Reply:
This condition checks whether a single expected workstatus exists or not. In the case that no expected workstatus was fetched at all, I still don't believe we should delete the entry here. We can instead have the reconciliation of an empty combinedstatus resolution delete the entry. What do you think?

Reply:
When the line I am commenting on is executed, are we guaranteed that the `csResolution` is empty? It could have been created and populated with content earlier, right?

Reply:
We are not guaranteed it is empty. Yes. In both cases that this whole call leads to a populated or empty combinedstatus, the combinedstatus identifier would be added to the set of identifiers to queue higher on the stack.

When syncing a combinedstatus (pending), an entry should be deleted if empty. This is what I meant above.